### ISSUE TITLE
stylelint-property-groups-structure

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#1575]
@evilebottnawi [asked](https://github.com/stylelint/stylelint/issues/1500#issuecomment-229985740):

> how about newlines between properties?

In [my original proposal](https://github.com/stylelint/stylelint/issues/1500#issuecomment-229300139) I spoke about a plugin, perhaps called `stylelint-property-groups-structure`:

 > Whereas, I think linting empty lines *before property groups* should be a plugin as there are all sorts of edge-cases in and around this feature. One option could be a plugin (perhaps called `stylelint-property-groups-structure`) that [builds on top of](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md#stylelintrules) `declaration-block-properties-order`. This kind of mixes the concerns of empty lines with ordering, but I can't see away around this in this very specific situation. I'm OK with the mixing of concerns in community-land, but only if it's absolutely necessary. 

An example config:

```js
[
  {
    emptyLineBefore: "always",
    order: "flexible",
    properties: [
      "color",
      "font-size",
      "font-weight",
    ],
  },
  {
    emptyLineBefore: "never",
    order: "flexible",
    properties: [
      "display",
      "position",
    ],
  },
]
```

The plugin checks for an empty line before each property group, it then strips the `emptyLineBefore` property from configuration object and passes the configuration object down to `declaration-block-properties-order` (which the plugin [wraps](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md#stylelintrules)) to do its stuff. This way you don't have to duplicate your property group config between two plugins/rules.

This is just a suggestion, but my initial thoughts are that this approach has mileage and is in-keeping with our division of plugins and rules.

**Comment by hudochenkov** [comment]
I think `stylelint-property-group-empty-line-before` will reflect true nature of this plugin.

**Comment by jeddy3** [comment]
> I think `stylelint-property-group-empty-line-before` will reflect true nature of this plugin.

Building on my explanation in https://github.com/stylelint/stylelint/issues/1500#issuecomment-230086153, I'll try to flesh this issue out a little in the hope of clarifying the nature of this plugin.

Unlike something like a declaration (`declaration-empty-line-before`) or an at-rule (`at-rule-empty-line-before`), a "property group" is a *user-defined* construct. Hence, we need to know what each property group looks like before we can check whether it is preceded by an empty line or not. So, the config would look something like:

```js
{
  "rules": {
    "property-group-empty-line-before": [
      {
        emptyLineBefore: true,
        properties: [
          "color",
          "font-size",
          "font-weight",
        ],
      },
      {
        emptyLineBefore: true,
        properties: [
          "display",
          "position",
        ],
      },
    ]
  }
}
```

This is fine and would work. However, the problem is that there is also the `declaration-block-properties-order` rule which needs to know about these user-defined property groups too, and *each rule is stand alone and doesn't know about the other rules*. We could duplicate the property group part of the configuration object:

```js
{
  "rules": {
    "declaration-block-properties-order": [
      {
        order: "flexible",
        properties: [
          "color",
          "font-size",
          "font-weight",
        ],
      },
      {
        order: "strict",
        properties: [
          "display",
          "position",
        ],
      },
    ]
  },
   "plugin/property-group-empty-line-before": [
      {
        emptyLineBefore: true,
        properties: [
          "color",
          "font-size",
          "font-weight",
        ],
      },
      {
        emptyLineBefore: true,
        properties: [
          "display",
          "position",
        ],
      },
    ]
  }
}
```

Another approach is... in the plugins system we have [the concept of](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md#stylelintrules) (I guess you could call them) "super" plugins that consume other rules and plugins.

So, we could create a plugin that wraps `declaration-block-properties-order` and augments it to *also check for empty lines before each property group*. I proposed that this be called `stylelint-property-groups-structure`, `property-groups` because it's specific to them and `structure` because it's concerned with both order *and* empty lines. This plugin would accept a "super" config object:

```js
{
  "rules": {
    "plugin/property-groups-structure": [
      {
        emptyLineBefore: true,
        order: "flexible",
        properties: [
          "color",
          "font-size",
          "font-weight",
        ],
      },
      {
        emptyLineBefore: true,
        order: "strict",
        properties: [
          "display",
          "position",
        ],
      },
    ]
  }
}
```

The internals of the plugin would roughly look like:

```js
export default stylelint.createPlugin("plugin/property-groups-structure", function (expectation) {
  return (root, result) => {
    // 1. First check that there is an empty line before each property group then
    // 2. Strip the config object so that it is compatible with `declaration-block-properties-order`
    //     by removing the `emptyLineBefore` from each object in the array e.g.
    const strippedConfigObject = delete expectation.emptyLineBefore;
    // 3. Run `declaration-block-properties-order` with this stripped config object e.g.
    const runPropertiesOrder = stylelint.rules["declaration-block-properties-order"](strippedConfigObject)
    runPropertiesOrder(root, result)
  }
})
```

As I said above:

> This kind of mixes the concerns of empty lines with ordering, but I can't see away around this **in this very specific situation**. I'm OK with the mixing of concerns in community-land, but only if it's absolutely necessary.

As an aside, I *don't* think it's absolutely necessary to mix concerns with [`block-non-properties-order`](https://github.com/stylelint/stylelint/issues/1500) and I explain why in https://github.com/stylelint/stylelint/issues/1500#issuecomment-230086153.

I think there are pros and cons to each approach:

`stylelint-property-group-empty-line-before`:

- Pro: Doesn't mix concerns, and perhaps doesn't start us down that slippery slope.
- Con: The user needs to duplicate their "property groups" config.

`stylelint-property-groups-structure`:

- Pro: The user *doesn't* need to duplicate their "property groups" config. 
- Con: Mixes concerns, and perhaps start us down a slippery slope that we have so far successfully avoided.

In case you're wondering why we don't just bundle all this together in one rule in core... we have [list of criteria](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#let-everyone-know) for including a rule in core. One of the criteria is that a rule must have "a clear and unambiguous finished state". Something like `declaration-block-properties-order` has a "a clear and unambiguous finished state", whereas things like `property-groups-structure`/`stylelint-property-group-empty-line-before` and `block-non-properties-order` do not as there will be new features to add (which would be breaking changes to the stylelint core) and there are plenty of edge-cases to consider. More background to this can be found in these two discussions: https://github.com/stylelint/stylelint/issues/1308#issuecomment-220208911, https://github.com/stylelint/stylelint/issues/1238#issuecomment-219236893.

Does the above help clear things up some what? If so, which approach do you think is best? We know of plugins that *conditionally* apply consumed rules, but I believe, if we went with the latter approach, this would be the first plugin to *extend* a consumed rule. Anyone apprehensive about setting that precedence?

**Comment by jeddy3** [comment]
If someone wants to experiment with this plugin, then please feel free to create a repo and have a a go. Also, feel free to at-mention me from there if you'd like to continue the discussion further.

**Comment by hudochenkov** [comment]
I'm thinking how to handle comments. They can be before the first group or between two groups. They can be on separate lines, or on a line with a declaration.

Given:

```js
{
  "rules": {
    "plugin/property-groups-structure": [
      {
        emptyLineBefore: "always",
        order: "strict",
        properties: [
          "border-bottom",
          "font-style",
        ],
      },
      {
        emptyLineBefore: "always",
        order: "strict",
        properties: [
          "display",
        ],
      },
      {
        emptyLineBefore: "always",
        order: "strict",
        properties: [
          "position",
        ],
      },
    ]
  }
}
```

The following example is not considered warnings:

```css
div {
  /* comment 1 */
  /* comment 2 */
  /* comment 3 */
  border-bottom: 1px solid red; /* inline comment 1 */
  font-style: italic;

  display: none; /* inline comment 2 */

  /* comment 1 */
  /* comment 2 */
  /* comment 3 */
  position: absolute;
}
```

I think comments before declaration should be connected to this declaration. And an inline comment should be connected to the previous declaration.

How to handle this dilemma? Is it ok to treat comments like in description above?

Also, this plugin should have an option like:

```js
{
  ignore: ["first-nested"]
}
```

Given:

```js
{
  "rules": {
    "plugin/property-groups-structure": [
      [
        {
          emptyLineBefore: "always",
          order: "strict",
          properties: [
            "display",
          ],
        },
        {
          emptyLineBefore: "always",
          order: "strict",
          properties: [
            "top",
          ],
        },
      ],
      {
        ignore: ["first-nested"],
      }
    ]
  }
}
```

The following pattern is considered warnings:

```css
a {

  top: 5px;
}
```

The following pattern is not considered warnings:

```css
a {
  top: 5px;
}
```

**Comment by jeddy3** [comment]
Firstly, thanks for making a start on this plugin too! You're on fire at the moment :)

Regarding you're two queries... I don't think there's much to do here.

> I'm thinking how to handle comments.

I think handling the empty space before comments is the responsibility of the [`comment-empty-line-before`](https://github.com/stylelint/stylelint/tree/master/src/rules/comment-empty-line-before) rule. 

And handling the empty space before declarations, when they are preceded by a comment, is done by the [`declaration-empty-line-before`](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-empty-line-before/README.md) rule and its [`"after`](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-empty-line-before/README.md#after-comment) `-` [`comment"`](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-empty-line-before/README.md#after-comment-1) options.

> Also, this plugin should have an option like: `ignore: ["first-nested"]`

I think this is also already covered by the [`declaration-empty-line-before`](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-empty-line-before/README.md)'s [`except: ["first-nested"]`](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-empty-line-before/README.md#first-nested) option _when combined_ with its new [`"ignore: ["after-declaration"]`](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-empty-line-before/README.md#after-declaration-1) option (which was added a few days ago).

So, I think, in this plugin, you should *only check* the empty line before a declaration if that declaration is preceded by another declaration i.e. *never check* the first declaration or those preceded by comments. Leave that to `declaration-empty-line-before`.

With the following config:

```js
{
  "rules": {
    "declaration-empty-line-before": ["always", {
      except: [
        "after-comment",
        "first-nested",
      ],
      ignore: [
        "after-declaration",
      ]
    }],
    "plugin/property-groups-structure": [
      {
        emptyLineBefore: "always",
        order: "strict",
        properties: [
          "display",
        ],
      },
      {
        emptyLineBefore: "always",
        order: "strict",
        properties: [
          "top",
        ],
      },
    ]
  }
}
```

These are all valid:

```css
a {
  display: block;
  
  top: 10px;
}
```

```css
a {
  --custom-prop: red;

  display: block;
  
  top: 10px;
}
```

```css
a {
  --custom-prop: red;

  display: block;
  
  /* comment */
  top: 10px;
}
```

I hope that makes sense :)

**Comment by hudochenkov** [comment]
> Firstly, thanks for making a start on this plugin too! You're on fire at the moment :)

Thank you :)

> So, I think, in this plugin, you should only check the empty line before a declaration if that declaration is preceded by another declaration...

I've got what you mean. Sounds good.

What about this case?

```js
{
  "rules": {
    "plugin/property-groups-structure": [
      {
        emptyLineBefore: "always",
        order: "strict",
        properties: [
          "display",
        ],
      },
      {
        emptyLineBefore: "always",
        order: "strict",
        properties: [
          "top",
        ],
      },
    ]
  }
}
```

```css
a {
  display: block; /* comment */

  top: 10px;
}
```

Is there any existing rule to specify empty line before `top`? Comment is related to `display`, and it's kinda should be ignored by the plugin.

**Comment by jeddy3** [comment]
> and it's kinda should be ignored by the plugin.

Yep, we call those "shared-line" comments and you'll just want to [ignore them](https://github.com/stylelint/stylelint/blob/master/src/rules/comment-empty-line-before/index.js#L60-L61).

So, when I say "declaration is preceded by another declaration". I mean preceded by _on another line_. So you'll need to skip over any "shared-line" comments you come across in the node tree when determining if a declaration is preceded by another declaration.

**Comment by hudochenkov** [comment]
Thank you :)

**Comment by hudochenkov** [comment]
Released and published to npm https://github.com/hudochenkov/stylelint-property-groups-structure

Thank you, @jeddy3 for all preparation work you did!

**Comment by evilebottnawi** [comment]
@hudochenkov In the near future I will start integration into their style may be suggestions for improvement :smile:

**Comment by jeddy3** [comment]
Well done. Congrats on another excellent plugin! :)

I'll create a PR adding it to the plugin list.

**Comment by hudochenkov** [comment]
@jeddy3 thank you :)
